How to Graph Structural Variations
=====================================



.. sourcecode:: r
    

    getwd()


::

    ## [1] "/Users/knagdimov/Documents/gTrack/rtdocs/vignetteSVs"


.. sourcecode:: r
    

    set


::

    ## Error in eval(expr, envir, enclos): object 'set' not found


.. sourcecode:: r
    

    ## this load sequences that have had coverage calculated from cancer cell lines (GRanges object, have to make into a gTrack)
    cov = readRDS('~/Documents/gTrack/inst/extdata/coverage.rds')
    
    ## wrap a gTrack around this, draw with blue circles, and label the track "Cov" and sets 0 as lower bound for all views 
    gt.cov = gTrack(cov, y.field = 'mean', circles = TRUE, col = 'blue', name = 'Cov')


::

    ## Error in eval(expr, envir, enclos): could not find function "gTrack"


.. sourcecode:: r
    

    ## this loads the junctions data from the cell line (GRangesList, where each item is a length 2 GRanges
    ## with strand information specifying the two locations and strands that are being fused) 
    junctions = readRDS('inst/extdata/junctions.rds')


::

    ## Warning in gzfile(file, "rb"): cannot open compressed file 'inst/extdata/
    ## junctions.rds', probable reason 'No such file or directory'



::

    ## Error in gzfile(file, "rb"): cannot open the connection


.. sourcecode:: r
    

    ## loading the GENCODE gene model gTrack (hg19 pre-loaded comes with gTrack,
    ## but can be made from any .gff file from GENCODE (http://www.gencodegenes.org/releases/19.html)
    gt.ge = track.gencode()


::

    ## Error in eval(expr, envir, enclos): could not find function "track.gencode"


.. sourcecode:: r
    

    ## this loads a gTrack object of a genome graph i.e. network of the same cancer cell line (generated by JaBba)
    graph = readRDS('inst/extdata/graph.rds')


::

    ## Warning in gzfile(file, "rb"): cannot open compressed file 'inst/extdata/
    ## graph.rds', probable reason 'No such file or directory'



::

    ## Error in gzfile(file, "rb"): cannot open the connection


.. sourcecode:: r
    

    ## pick an interesting junction and plot the genes, coverage, and genome graph around it
    ## the links argument specifies the junctions that are being drawn
    window = junctions[[290]] + 1e5




.. sourcecode:: r
    

    plot(c(gt.ge, gt.cov, graph), window, links = junctions)


::

    ## Error in plot(c(gt.ge, gt.cov, graph), window, links = junctions): object 'gt.ge' not found


