How to Graph Structural Variations
=====================================



.. sourcecode:: r
    

    ## make sure to load in these libraries
    library(gTrack)



.. sourcecode:: r
    

    options(warn=-1)
    ## this load sequences that have had coverage calculated from cancer cell lines (GRanges object, have to make into a gTrack)
    cov = readRDS('inst/extdata/coverage.rds')


::

    ## Error in gzfile(file, "rb"): cannot open the connection


.. sourcecode:: r
    

    ## wrap a gTrack around this, draw with blue circles, and label the track "Cov" and sets 0 as lower bound for all views 
    gt.cov = gTrack(cov, y.field = 'mean', circles = TRUE, col = 'blue', name = 'Cov')


::

    ## Error in .local(.Object, ...): check input: gTrack objects can only be defined around GRanges, GRangesLists, RleLists, ffTrack, file paths to .rds files of the latter object types, or file paths to  UCSC format files


.. sourcecode:: r
    

    ## this loads the junctions data from the cell line (GRangesList, where each item is a length 2 GRanges
    ## with strand information specifying the two locations and strands that are being fused) 
    junctions = readRDS('inst/extdata/junctions.rds')


::

    ## Error in gzfile(file, "rb"): cannot open the connection


.. sourcecode:: r
    

    ## loading the GENCODE gene model gTrack (hg19 pre-loaded comes with gTrack,
    ## but can be made from any .gff file from GENCODE (http://www.gencodegenes.org/releases/19.html)
    gt.ge = track.gencode()


::

    ## Pulling gencode annotations from /Library/Frameworks/R.framework/Versions/3.3/Resources/library/gTrack/extdata/gencode.composite.collapsed.rds


.. sourcecode:: r
    

    ## this loads a gTrack object of a genome graph i.e. network of the same cancer cell line (generated by JaBba)
    graph = readRDS('inst/extdata/graph.rds')


::

    ## Error in gzfile(file, "rb"): cannot open the connection


.. sourcecode:: r
    

    ## pick an interesting junction and plot the genes, coverage, and genome graph around it
    ## the links argument specifies the junctions that are being drawn
    window = junctions[[290]] + 1e5


::

    ## Error in eval(expr, envir, enclos): object 'junctions' not found




.. sourcecode:: r
    

    plot(c(gt.ge, gt.cov, graph), window, links = junctions)


::

    ## Error in plot(c(gt.ge, gt.cov, graph), window, links = junctions): object 'gt.cov' not found




.. sourcecode:: r
    

    ## Load in coverage data for the HCC1143 cell line 
     ocovh = readRDS('inst/extdata/10X/files/HCC1143/cov/cov.rds')
    
     cov = gr.tile(seqlengths(ocovh, 5e3)
     cov = readRDS("inst/extdata/segstats_cov.rds")


::

    ## Error: <text>:5:2: unexpected symbol
    ## 4:  cov = gr.tile(seqlengths(ocovh, 5e3)
    ## 5:  cov
    ##     ^


