How to Graph Structural Variations
=====================================


.. sourcecode:: r
    

    ## this load sequences that have had coverage calculated from cancer cell lines (GRanges object, have to make into a gTrack)
    cov = readRDS('~/Documents/gTrack/inst/extdata/coverage.rds')
    
    ## wrap a gTrack around this, draw with blue circles, and label the track "Cov" and sets 0 as lower bound for all views 
    gt.cov = gTrack(cov, y.field = 'mean', circles = TRUE, col = 'blue', name = 'Cov')


::

    ## Warning in `[<-`(`*tmp*`, null.ix, value = list(<S4 object of class
    ## structure("GRanges", package = "GenomicRanges")>)): implicit list embedding
    ## of S4 objects is deprecated


.. sourcecode:: r
    

    ## this loads the junctions data from the cell line (GRangesList, where each item is a length 2 GRanges
    ## with strand information specifying the two locations and strands that are being fused) 
    junctions = readRDS('~/Documents/gTrack/inst/extdata/junctions.rds')
    
    ## loading the GENCODE gene model gTrack (hg19 pre-loaded comes with gTrack,
    ## but can be made from any .gff file from GENCODE (http://www.gencodegenes.org/releases/19.html)
    gt.ge = track.gencode()


::

    ## Pulling gencode annotations from /Library/Frameworks/R.framework/Versions/3.3/Resources/library/gTrack/extdata/gencode.composite.collapsed.rds


.. sourcecode:: r
    

    ## this loads a gTrack object of a genome graph i.e. network of the same cancer cell line (generated by JaBba)
    graph = readRDS('~/Documents/gTrack/inst/extdata/graph.rds')
    
    ## pick an interesting junction and plot the genes, coverage, and genome graph around it
    ## the links argument specifies the junctions that are being drawn
    window = junctions[[290]] + 1e5




.. sourcecode:: r
    

    plot(c(gt.ge, gt.cov, graph), window, links = junctions)


::

    ## Warning in `[<-`(`*tmp*`, null.ix, value = list(<S4 object of class
    ## structure("GRangesList", package = "GenomicRanges")>, : implicit list
    ## embedding of S4 objects is deprecated



::

    ## Warning in gr.findoverlaps(query, subject, ...): seqlength mismatch .. no
    ## worries, just letting you know

    ## Warning in gr.findoverlaps(query, subject, ...): seqlength mismatch .. no
    ## worries, just letting you know



::

    ## Warning in gr.findoverlaps(gr, windows): seqlength mismatch .. no worries,
    ## just letting you know


.. figure:: figure/plot-firstSV-1.png
    :alt: plot of chunk plot-firstSV

    plot of chunk plot-firstSV
